sdk/client/src/verifying_client.rs

```diff
@@ -0,0 +1,516 @@
+            .await?
+            .into_parts();
+        let state_proof: StateProof = (&state_proof_view).try_into()?;
+
+        let latest_li = state_proof.0.ledger_info();
+        ensure!(latest_li.version() == state.version);
+        ensure!(latest_li.timestamp_usecs() == state.timestamp_usecs);
+
+        self.verify_and_ratchet(&state_proof)?;
+
+        Ok(state_proof.1.more)
+    }
+
+    /// Verify and ratchet forward our trusted state using a state proof.
+    pub fn verify_and_ratchet(&self, state_proof: &StateProof) -> Result<()> {
+        let (latest_li, epoch_change_proof, _) = state_proof;
+
+        let trusted_state = self.trusted_state();
+        let change = trusted_state.verify_and_ratchet(latest_li, epoch_change_proof)?;
+
+        if let Some(new_state) = change.new_state() {
+            self.trusted_state_store
+                .write()
+                .unwrap()
+                .ratchet(new_state)?;
+        }
+
+        // TODO(philiphayes): verify the accumulator consistency proof; no-one
+        // seems to actually verify this anywhere?
+
+        Ok(())
+    }
+
+    pub async fn get_account(
+        &self,
+        address: AccountAddress,
+    ) -> Result<Response<Option<AccountView>>> {
+        Ok(self
+            .send_via_batch(MethodRequest::GetAccount((address,)))
+            .await?
+            .and_then(MethodResponse::try_into_get_account)?)
+    }
+
+    /// Send a single request via `VerifyingClient::batch`.
+    async fn send_via_batch(&self, request: MethodRequest) -> Result<Response<MethodResponse>> {
+        let mut responses = self.batch(vec![request]).await?.into_iter();
+        let response = match responses.next() {
+            Some(response) => response,
+            None => bail!("response batch is empty"),
+        };
+        ensure!(responses.next().is_none(), "too many responses");
+        response
+    }
+
+    pub async fn batch(
+        &self,
+        requests: Vec<MethodRequest>,
+    ) -> Result<Vec<Result<Response<MethodResponse>>>> {
+        // transform each request into verifying sub-request batches
+        let batch = VerifyingBatch::from_batch(requests);
+
+        // flatten and collect sub-request batches into flat list of requests
+        let mut requests = batch.collect_requests();
+        // append get_state_proof request
+        let start_version = self.version();
+        requests.push(MethodRequest::get_state_proof(start_version));
+
+        // actually send the batch
+        let responses = self.inner.batch(requests).await?;
+
+        // TODO(philiphayes): REMOVE. for now, just convert the errors until we
+        // use the real error type.
+        let mut responses = responses
+            .into_iter()
+            .map(|result| result.map_err(anyhow::Error::new))
+            .collect::<Vec<_>>();
+
+        // extract and verify the state proof
+        let (state_proof_response, state) = responses.pop().unwrap()?.into_parts();
```

Comment:
should rely on request id to get state_proof_response, the order of responses are not guaranteed to be same with requests order.

Reply:
Actually the rust sdk's batching function guarantee that the order of responses matches the order of the requests. See https://github.com/diem/diem/blob/main/sdk/client/src/lib.rs#L100

Reply:
It seems json-rpc services aren't guaranteed to return responses in-order (in general), but our specific Diem json-rpc service does guarantee this. At this layer (wrapping the result from the inner `Client`), we should be fine assuming the correct ordering. If we want to be super compliant, I suppose `Client` could guarantee order

Reply:
For reference, here's how we're currently handling request batches in the json-rpc service:

https://github.com/diem/diem/blob/main/json-rpc/src/runtime.rs#L307

```rust
                // batch API call
                let futures = requests.into_iter().map(|req| {
                    rpc_request_handler(
                        req,
                        service.clone(),
                        Arc::clone(&registry),
                        ledger_info.clone(),
                        LABEL_BATCH,
                        trace_id.clone(),
                        sdk_info,
                    )
                });
                let responses = join_all(futures).await;
```

https://docs.rs/futures/0.3.14/futures/future/fn.join_all.html

> The returned future will drive execution for all of its underlying futures, collecting the results into a destination `Vec<T>` **in the same order as they were provided.**